Add support for GPX tracks with multiple nested track segments

bjohare opened the pull request:
### Proposed changes in this pull request

#### Fixes #1350: Some specific formats for GPX files are failing

- Removed `GPX` parsing with `GDAL` and replaced with `gpxpy`
- Added test for parsing of multiple nested `<trkseg>` elements within a single `<trk>` element
- Added sample `GPX` file with nested `<trkseg>` elements
- Updated requirements
- Refactored some exception handling in `GPXProcessor` and `resources/models.py`

### When should this PR be merged

Can be merged anytime.

### Risks

Only `gpx` files with xml versions `1.0` or `1.1` are supported. Some devices (incorrectly) produce xml version 2.0 which doesn't exist. Depending on the device used to produce the xml there may need to be some pre-processing of the resources to force compliance with the xml spec.

### Follow-up actions

None required

### Checklist (for reviewing)

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?** 
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

alukach commented on cadasta/resources/processors/gpx.py:
```diff
@@ -1,19 +1,80 @@
+            multiline += parse_routes(self.gpx.routes)
+            layers['routes'] = GeometryCollection(
+                MultiLineString(multiline)
+            )
```
Is it possible for a `.gpx` file to have _both_ tracks and routes?  If so, it appears that the tracks data will appear in the `'routes'` layer.  Is this desired?

bjohare replied:
Yeah, 'tis a bug. Have fixed in my latest commit.

alukach commented on cadasta/resources/processors/gpx.py:
```diff
@@ -1,19 +1,80 @@
+            )
+        if self.gpx.waypoints:
+            multipoint = []
+            multipoint += parse_waypoints(self.gpx.waypoints)
```
Is there any reason to do the `multipoint = []; multipoint += parse_waypoints(self.gpx.waypoints)` rather that just `multipoint = parse_waypoints(self.gpx.waypoints)`?  Or even:

```python
layers['waypoints'] = GeometryCollection(
    MultiPoint(parse_waypoints(self.gpx.waypoints))
)
```

... or maybe that's _too_ dense? 🤷‍♂️

bjohare replied:
Have tidied this up a bit..

bjohare commented:
@alukach thanks for the review. Have added some tests to confirm that files containing both `<rte>` and `<trk>` elements get parsed correctly. Also made some tweaks to `map_utils.js` to take account of pagination work and to make sure that multiple layers per single resource get displayed properly.

alukach commented:
@bjohare Fixes look good 👌 . What's the backstory on `map_utils.js`? I'm guessing that was a bug introduced by #1425 ? What endpoint does it request from? I'm not seeing how pagination is followed (at least in the changes you made), perhaps that's not what you added?

bjohare commented:
@alukach a javascript callback broke in `map_utils.js` because the format of the response to the spatialunit api endpoint had changed. That endpoint is at `api/v1/organizations/{org}/projects/{prj}/spatialresources`. It should really be an async endpoint but was implemented before we had async views. I'm not following pagination there yet. I'll open a ticket to convert the endpoint from `api` to `async`.

alukach commented:
Okay, thanks for clearing that up. So, definitely a bug if we're not following pagination on that view.

> It should really be an async endpoint but was implemented before we had async views. I'm not following pagination there yet.

I'm not really aware of what the distinctions are between the two, however maybe I'll save my questions for the issue you create related to this topic.

bjohare commented on cadasta/resources/models.py:
```diff
@@ -30,6 +30,8 @@
 
 GPX_MIME_TYPES = ('application/xml', 'text/xml', 'application/gpx+xml')
 
+logger = logging.getLogger(__name__)
```
Good call. I'll remove that.

bjohare commented on cadasta/resources/models.py:
```diff
@@ -192,17 +194,13 @@ def create_spatial_resource(sender, instance, created, **kwargs):
                     if len(layers[layer]) > 0:
                         SpatialResource.objects.create(
                             resource=instance, name=layer, geom=layers[layer])
+                os.remove(write_path)
```
If the `InvalidGPXFile` exception is raised then the temporary file will not be removed.

seav replied:
Ah, right. Ignore this comment then.

alukach replied:
I hadn't paid much attention to the aspect of saving temporary files to disk when I reviewed this PR earlier.  This current implementation feels a bit error-prone.  If an exception is thrown in between `with open(write_path, 'wb') as f:` and `os.remove(write_path)`, the path will never be deleted.  It sounds like the servers are configured with a cronjob to clear out temp files every so often, however I think we could greatly minimize risk around the volume of saved files with one of two patterns:

a) Use `try/finally`:

```python
with open(write_path, 'wb') as f:
    f.write(file)
try:
    if mime_type in GPX_MIME_TYPES:
        # ...
    else:
        raise InvalidGPXFile(
            # ...
        )
finally:
    os.remove(write_path)
```

b) Use a context manager:

```python
import tempfile

# Create a tempfile with visible path on FS (accessible via f.name)
with tempfile.NamedTemporaryFile() as f:
    f.write(instance.file.open().read())

    # need to double check the mime-type here as browser detection
    # of gpx mime type is not reliable
    mime = magic.Magic(mime=True)
    mime_type = str(mime.from_file(f.name), 'utf-8')

    if mime_type in GPX_MIME_TYPES:
        processor = GPXProcessor(f.name)
        layers = processor.get_layers()
        for layer in layers.keys():
            if len(layers[layer]) > 0:
                SpatialResource.objects.create(
                    resource=instance, name=layer, geom=layers[layer])
    else:
        raise InvalidGPXFile(
            _("Invalid GPX mime type: {error}".format(
                error=mime_type))
        )
```

If using a collection of files, something like `tempfile.TemporaryDirectory` makes it easy to get a directory to use for larger operations.

With the context manager, it frees the developer from having to think about the cleanup.  In either of the above examples, we can have high level of confidence that the file will be removed (save for a massive failure on our system).

Ultimately, I think the general idea of `os.path.join(settings.MEDIA_ROOT, 'temp', 'somefile')` is a bit of an anti=pattern as it does not guarantee cleanup at the end of the request/response cycle. This PR didn't introduce that code and that pattern exists throughout the Cadasta Platform codebase so I'm not hung up on it being fixed in this PR.  However, I do think we should start moving towards stricter tempfile management.

bjohare replied:
@alukach I agree with you about our handling of temporary files. I've implemented the second pattern in my latest changes. The `NamedTemporaryFile` is now passed the path to the temporary directory created by the call to [`io.ensure_dirs()`](https://github.com/Cadasta/cadasta-platform/blob/master/cadasta/resources/utils/io.py#L5).

bjohare commented on cadasta/resources/processors/gpx.py:
```diff
@@ -1,19 +1,75 @@
-            if name in KEEP_LAYERS:
-                # geom = self._get_features(layer)
-                layers[name] = GeometryCollection(layer.get_geoms(geos=True))
+        if self.gpx.tracks:
```
I think its better to reflect the structure of the `GPX` file. The user is then able to switch these layers on/off separately in the UI.

seav replied:
I understand. I guess my only concern is that I think our typical users would not necessarily know the technical details of how standard GPX files are actually structured. My personal preference from a UI/UX point of view is to toggle individual GPX resource files (so everything in a GPX file—waypoints, routes, tracks—are toggled together) instead of toggling per layer. I think it makes things simpler for our users and would also make the Leaflet layer control shorter.

bjohare replied:
I think we need to review how this is done. I think we should be moving towards a more customized UI for adding map layers maybe as part of https://github.com/Cadasta/cadasta-platform/issues/1476. The grouped layer control has additional issues in that we can't paginate it if there are lots of gpx resources. My feeling is that this is ok for now. If we need to make changes to the grouped layer control then we should probably open a separate issue?

seav replied:
I agree. Let's revisit this in the future.

bjohare commented on cadasta/resources/processors/gpx.py:
```diff
@@ -1,19 +1,75 @@
+            )
+        if not layers:
+            raise InvalidGPXFile(
+                _('Error parsing GPX file: no geometry found.'))
```
Ok I'll fix these..

seav replied:
This particular line is not yet addressed.

bjohare replied:
Oops!

bjohare commented on cadasta/resources/processors/gpx.py:
```diff
@@ -1,19 +1,75 @@
+    multiline = []
+    for track in tracks:
+        for segment in track.segments:
+            track_list_of_points = parse_segment(segment)
```
yep, will change..

bjohare commented on cadasta/resources/processors/gpx.py:
```diff
@@ -1,19 +1,75 @@
+
+
+def parse_segment(segment):
+    track_list_of_points = []
```
You're right. I'll refactor this. I think I'll leave the `parse_segment` function though rather than repeat the list comprehension in two places further down..

seav commented on cadasta/resources/processors/gpx.py:
```diff
@@ -1,19 +1,75 @@
+
+
+def parse_waypoints(waypoints):
+    multipoint = []
```
This feedback item is still not yet addressed.

bjohare replied:
Ok, thanks.

bjohare commented:
@alukach  and @seav it would be good to get a review of the latest commit on this branch. There were some issues with `gpxpy` logging which I've addressed. I've overridden `gpxpy` parser to prevent it logging exceptions. Also added a test for the new parser class.

alukach commented:
@bjohare sure thing, can you describe the backstory? Why shouldn't the parser log exceptions? How was this discovered?

bjohare commented:
@alukach I'm still not entirely clear what was causing the problem. Python logging was choking on non-unicode data generated when an invalid GPX file was uploaded, but why it should encounter ascii data is a bit of a mystery. The error happened when a `logging.exception()` call was encountered in the `parser` exception handler. Simplest solution was to override the `parse` method.

seav commented:
The changes look OK. I really don't like that we are overriding the `parse` method in order to prevent logging but I guess it can't be helped.

bjohare commented:
@seav I hear you, I'm not sure what else I can do at the moment though. I'm open to suggestions!

seav commented:
If we are not using logging at all anywhere, I wonder if we can just disable all logging entirely by setting the logging level to a really high integer: `logging.disable(100)`

I think one of the problems is that gpxpy uses the root logger for logging messages. This prevents users of the module from disabling logging just for gpxpy. Somebody actually filed an issue for this but the maintainer closed it because they are looking for a PR: https://github.com/tkrajina/gpxpy/issues/79

bjohare commented:
We log invalid odk submissions to help with debugging.. https://github.com/Cadasta/cadasta-platform/blob/master/cadasta/xforms/views/api.py#L52.

bjohare commented:
@alukach are you happy for this to be merged?

alukach commented:
@bjohare Looks good to me, thanks for talking me through the reasoning.
